torchrl/envs/utils.py

```diff
@@ -181,34 +187,46 @@ def step_mdp(
             next_tensordict.update(out)
             return next_tensordict
         return out
-    out = tensordict.get("next").clone(False)
-    excluded = set()
-    if exclude_done:
-        excluded.add(done_key)
-    if exclude_reward:
-        excluded.add(reward_key)
-    if len(excluded):
-        out = out.exclude(*excluded, inplace=True)
-    td_keys = None
+
+    td = tensordict.exclude("next")
+    td_next = tensordict.get("next")
+
+    td_keys = td.keys(True, True)
+    td_next_keys = td_next.keys(True, True)
+
+    out = TensorDict({}, td_next.batch_size)
+
+    # Set the keys from next
+    excluded = {
+        done_key if exclude_done else None,
+        reward_key if exclude_reward else None,
+    }
+    for key in td_next_keys:
+        if key not in excluded:
+            _set_key(dest=out, source=td_next, key=key)
+
+    # Set the keys from root
+    if not exclude_action:
+        _set_key(dest=out, source=tensordict, key=action_key)
```

Comment:
here we do the action separate just to not have another if later, perf is the same